src/main/java/org/opentripplanner/graph_builder/module/osm/naming/SidewalkNamer.java

```diff
@@ -0,0 +1,275 @@
+import org.locationtech.jts.operation.distance.DistanceOp;
+import org.opentripplanner.framework.geometry.GeometryUtils;
+import org.opentripplanner.framework.geometry.HashGridSpatialIndex;
+import org.opentripplanner.framework.geometry.SphericalDistanceLibrary;
+import org.opentripplanner.framework.i18n.I18NString;
+import org.opentripplanner.framework.lang.DoubleUtils;
+import org.opentripplanner.framework.logging.ProgressTracker;
+import org.opentripplanner.graph_builder.services.osm.EdgeNamer;
+import org.opentripplanner.openstreetmap.model.OSMWithTags;
+import org.opentripplanner.street.model.edge.StreetEdge;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * A namer that assigns names of nearby streets to sidewalks if they meet certain
+ * geometric similarity criteria.
+ * <p>
+ * The algorithm works as follows:
+ *  - for each sidewalk we look up (named) street edges nearby
+ *  - group those edges into groups where each edge has the same name
+ *  - draw a flat-capped buffer around the sidewalk, like this: https://tinyurl.com/4fpe882h
+ *  - check how much of a named edge group is inside the buffer
+ *  - remove those groups which are below MIN_PERCENT_IN_BUFFER
+ *  - take the group that has the highest percentage (as a proportion of the sidewalk length) inside
+ *    the buffer and apply its name to the sidewalk.
+ * <p>
+ * This works very well for OSM data where the sidewalk runs a parallel to the street and at each
+ * intersection the sidewalk is also split. It doesn't work well for sidewalks that go around
+ * the corner, like https://www.openstreetmap.org/way/1059101564. These cases are, however, detected
+ * by the above algorithm and the sidewalk name remains the same.
+ */
+public class SidewalkNamer implements EdgeNamer {
+
+  private static final Logger LOG = LoggerFactory.getLogger(SidewalkNamer.class);
+  private static final int MAX_DISTANCE_TO_SIDEWALK = 50;
+  private static final double MIN_PERCENT_IN_BUFFER = .85;
+  private static final int BUFFER_METERS = 25;
+
+  private HashGridSpatialIndex<EdgeOnLevel> streetEdges = new HashGridSpatialIndex<>();
+  private Collection<EdgeOnLevel> unnamedSidewalks = new ArrayList<>();
+
+  @Override
+  public I18NString name(OSMWithTags way) {
+    return way.getAssumedName();
+  }
+
+  @Override
+  public void recordEdge(OSMWithTags way, StreetEdge edge) {
+    if (way.isSidewalk() && way.needsFallbackName() && !way.isExplicitlyUnnamed()) {
+      unnamedSidewalks.add(new EdgeOnLevel(edge, way.getLevels()));
+    }
+    if (way.isNamed() && !way.isLink()) {
+      // we generate two edges for each osm way: one there and one back. since we don't do any routing
+      // in this class we don't need the two directions and keep only one of them.
+      var containsReverse = streetEdges
+        .query(edge.getGeometry().getEnvelopeInternal())
+        .stream()
+        .anyMatch(candidate -> candidate.edge.isReverseOf(edge));
+      if (!containsReverse) {
```

Comment:
What do you think about changing the calling code so that it always sends pairs of street edges so that we don't have to do the lookup in the index?

Reply:
Considering that both edges were already grouped together in a Record in the caller, this looks like a good idea. It seems appropriate for the `EdgeNamer#recordEdge` function to be called with a way and both edges derived from it so "more complicated names can be built in the post-processing step". 

I'm slightly hesitant about the changes this pushes down into the PortlandCustomNamer implementation because it's creating a large number of new ArrayLists and Iterators and repeating the way tag parsing logic twice for each pair. But I don't immediately see a clean way to factor out the way tag parsing part. If you're not seeing any performance impact from this it seems fine.

Reply:
I would say that allocating a few extra collections is a price well worth paying for the added clarity.

The repeated parsing of the way properties happened even before this PR. Nevertheless, I have optimized this part of the Portland namer in e28a69c75fd5d4681f2f8a6213caba640671327c. Does this alleviate your concerns?

Reply:
As I mentioned, I didn't see a simple way to improve the situation, so if there's no measurable performance hit I'm fine with however you choose to implement it. In any case, the way you factored out isHighwayLink is nice and readable. Looks fine to me.